posthog/hogql_queries/insights/trends/calendar_heatmap_query_runner.py

```diff
@@ -0,0 +1,334 @@
+from datetime import timedelta, datetime
+from math import ceil
+from typing import Any, Optional, Union
+
+from django.db import models
+from django.db.models.functions import Coalesce
+
+from posthog.caching.insights_api import (
+    BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL,
+    REAL_TIME_INSIGHT_REFRESH_INTERVAL,
+    REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL,
+)
+from posthog.hogql import ast
+from posthog.hogql.constants import LimitContext
+from posthog.hogql.parser import parse_select
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql.property import action_to_expr, property_to_expr
+from posthog.hogql.timings import HogQLTimings
+from posthog.hogql_queries.insights.trends.series_with_extras import SeriesWithExtras
+from posthog.hogql_queries.query_runner import QueryRunner
+from posthog.hogql_queries.utils.query_date_range import QueryDateRange
+from posthog.models import Team
+from posthog.models.action.action import Action
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.property_definition import PropertyDefinition
+from posthog.schema import (
+    ActionConversionGoal,
+    ActionsNode,
+    CachedCalendarHeatmapQueryResponse,
+    CustomEventConversionGoal,
+    DashboardFilter,
+    DataWarehouseNode,
+    EventsNode,
+    HogQLQueryModifiers,
+    IntervalType,
+    CalendarHeatmapQuery,
+    CalendarHeatmapResponse,
+    EventsHeatMapDataResult,
+    EventsHeatMapRowAggregationResult,
+    EventsHeatMapColumnAggregationResult,
+    EventsHeatMapStructuredResult,
+)
+
+SEPARATOR = "','"
+
+templateUniqueUsers = """
+SELECT
+    mapKeys(query.hoursAndDays) as hoursAndDaysKeys,
+    mapValues(query.hoursAndDays) as hoursAndDaysValues,
+    mapKeys(query.hours) as hoursKeys,
+    mapValues(query.hours) as hoursValues,
+    mapKeys(query.days) as daysKeys,
+    mapValues(query.days) as daysValues,
+    query.total
+FROM (
+    SELECT
+        uniqMap(map(concatWithSeparator({separator},toString(toDayOfWeek(uniqueSessionEvents.timestamp)),toString(toHour(uniqueSessionEvents.timestamp))), uniqueSessionEvents.person_id)) as hoursAndDays,
+        uniqMap(map(toHour(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as hours,
+        uniqMap(map(toDayOfWeek(uniqueSessionEvents.timestamp), uniqueSessionEvents.person_id)) as days,
+        uniq(person_id) as total
+    FROM (
+        SELECT
+            any(events.person_id) as person_id,
+            session.session_id as session_id,
+            min(session.$start_timestamp) as timestamp
+        FROM events
+        WHERE and(
+            {event_expr},
+            {all_properties},
+            {test_account_filters}
+        )
+        GROUP BY session_id
+    ) as uniqueSessionEvents
+    WHERE {current_period}
+) as query
```

Comment:
I think the session's join here will only hold true for the context of WebAnalytics/SessionReplay. It is flipping the problem you had for the WebOverview the other way around: it won't match the `math` for `unique users` using only events like others insights will do.

e.g: if people have a "Big Number" tile in a dashboard for the count of unique users, the `query.total` here won't match.

IMHO it is better to count using only events for the time being to keep it more performant and simple and maybe add a `countUsingSessions` options to keep the behaviour for web analytics on another PR.

Reply:
+1 to this